### DISCUSSION TITLE
Creating a query serialiser for ORM async queries?

### DISCUSSION THREAD

**Discussion by fleetofthought** [sqlalchemy/sqlalchemy#7393] [Usage Questions]
Using SQLAlchemy async ORM 1.4, Postgres 12, Python 3.7

I am using an `augmented Declarative Base` with the SA ORM. I am using SA's `automap` extension because the database isn't
set up with models.py, rather a json file with the db schema gets parsed and committed directly to the db.

I will give a very simple example of my class, a read function and an update function.
```py

class Db:

    def __init__(self, config):
        self._engine = create_async_engine('postgresql+asyncpg://postgres@localhost:5432/db', echo=True, future=True)
        self._session = sessionmaker(self._engine, expire_on_commit=False, class_=AsyncSession)
        self._meta = MetaData()
        self._cache = self.cache()

    @property
    def meta(self):
        self._meta.reflect(bind=sync_engine)
        return self._meta

    @property
    def cache(self):
        Base = automap_base()
        Base.prepare(sync_engine, reflect=True)
        return Base

    # Simple read with an aggregate function
    async def read(self, collection, mods):

    Table = self._cache.classes[collection]

    async with self._session() as s:

        matches = set(Table.__mapper__.columns.keys()) & set(mods['fields'])
        for m in matches:
            q = select(func.sum(getattr(Table, m)))

            result = await session.execute(q)

            res = result.all()
            for row in res:
                print(row)

    # Update a single field
    async def updater(self, collection, query, data):

       Table = self._cache.classes[collection]

       async with self._session() as s:
           q = update(Table).filter_by(**query)
           q = q.values(**data)

           await s.execute(q)
           await s.commit()

db = Db()
asyncio.run(db.read('user', {'fields': ['usage', 'allowance']}))
asyncio.run(db.updater('user', {'name': 'Cheshire'}, {'org': 'Wonderland'}))
```

I am trying to understand how I can increase the query complexity.

For example, if we want a SQL query that looks something like:

```
UPDATE appuser 
SET org = 'Wonderland', role = '42'  
WHERE (name = 'Alice' OR name = 'Cheshire') 
AND (id > 0);
```

And we call the function with parameters like:
```
asyncio.run(updater('user',
                {'name': 'Alice'},
                {'or': [{'name': 'Cheshire'}]},
                {'and': [{'id': '> 0'}]},
                {'values': ['org': 'Wonderland', 'role': 42 ]}))
```
I have tried many variations of trying to chain select, query, or filter_by expressions together and I keep receiving errors like:

```
query = update(Table).filter_by(**where_condition, **or_condition)
query = update(Table).filter_by(**where_condition).filter_by( **or_condition)
query = update(Table).filter(**where_condition).filter( **or_condition)
Errors: sqlalchemy.exc.InvalidRequestError: Entity namespace for user" has no property "or"
```
```
if 'or' in or_condition:
    query = query.filter_by(**or_condition['or'])
Errors: TypeError: filter_by() argument after ** must be a mapping, not list
```
```
Select has no query method`
```
Using `with_loader_criteria` also wasn't a success.

These errors make sense, but I'm not sure how to get past them.

I am reading through the documentation
https://docs.sqlalchemy.org/en/14/core/tutorial.html
https://docs.sqlalchemy.org/en/14/tutorial/index.html
https://docs.sqlalchemy.org/en/14/orm/loading_objects.html

But am still really unclear between ORM and Core syntax and how they interact. There are instances where I begin a query
with the new ORM syntax but then need an operation that needs to be done through Core. When I try to chain the Core syntax
to the ORM syntax there are frequent errors.
This makes sense, but I am not sure how to integrate the two to achieve full functionality.

Does anyone have an example of chaining together filters by _parsing the function's input parameter dicts_ (not hardcoding the data in the function) so we can build up complex queries as we move through a long `read` or `update` function?

Perhaps there are examples somewhere that I missed, I would be happy to be pointed in the right direction!

**Comment by CaselIT** [comment]
Hi,

If I understand correctly you don't want to use sqlalchemy in the rest of the application, but you want to offer an interface on top of it.

For your question regarding the or, `filter_by` works only with and. To pass a complex clause you can use [`and_`](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.and_) and [`or_`](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.or_) with [`filter/where`](https://docs.sqlalchemy.org/en/14/core/selectable.html#sqlalchemy.sql.expression.Select.filter).

For example you can do `query.filter(sa.or_(sa.and_(table.c.id > 1, table.c.name != 'foo'), table.c.name=='bar'))`.

**Reply by fleetofthought** [reply]
Hi,

Thanks @CaselIT  for the response!
 
1. For your answer where you give the example` query.filter(sa.or_(sa.and_(table.c.id > 1, table.c.name != 'foo'), table.c.name=='bar')).`
Is this using the ORM or Core?

2. I may have been using some incorrect terminology for my initial question, I am going to try to clarify:
I very much would like to use the SA ORM in my code! That is actually what I am trying to achieve, I might be going about it in the wrong way.

When I said "serialiser" I meant I am building a class that I should be able to take in any input for a CRUD operation and output the correct SQL query using the ORM. This is why building up the query bit by bit, as the functions work on the input, is important for this use case.

Since I don't have fully working code yet, I am going to write some pseudo-code by way of example.
 
Let's say there's a `read` function that takes in these parameters when called: 

read('user', 
  {'status': 'open', 'or': 'closed'},
  {'fields':
      [
          'apps_used.name',
          {'max': 'time_started'},
          {'max': 'time_ended'},
          {'median': 'time_spent_online'},
          {'count': 'logins'},
      ],
      'group_by': ['apps_used.name'],
      'join': [{'from': 'apps_used.id', 'to': 'apps_used.id', 'fields': []}],
  }
  )

We would want to parse all these fields and gradually build up a query that we would then execute on the database. Here's some pesudo code, including the class initialization for clarity:

class Db:

    def __init__(self, config):
        self._engine = create_async_engine('postgresql+asyncpg://postgres@localhost:5432/db', echo=True, future=True)
        self._session = sessionmaker(self._engine, expire_on_commit=False, class_=AsyncSession)
        self._meta = MetaData()
        self._cache = self.cache()

    @property
    def meta(self):
        self._meta.reflect(bind=sync_engine)
        return self._meta

    @property
    def cache(self):
        Base = automap_base()
        Base.prepare(sync_engine, reflect=True)
        return Base

    async def read(self, table, query, fields):

        Table = self._cache.classes[table]

        async with self._session() as s:
            q = select(Table)   # Start building up the query with this select statement
            if query:
                if q['or']:
                    q = q.filter(or_(q['or'])  # Unpack the where condition into the query
                else:
                    q = q.where(**query)  # Unpack the where condition into the query

            if fields:
                 q = q.filter_by(fields[0])  # This is the column to get aggregates on
                 for i in fields['fields']:
                    if 'max' in i:
                        q = q.filter(func.max(**i))  # Unpack all instances of 'i' i.e.: {'max': 'time_started'},
                                                     # {'max': 'time_ended'} to get the 'max' value of these fields
                    if 'median' in i:
                        q = q.filter(func.percentile_cont(0.5).within_group(fields[0]))
                    
                    if 'count' in i:
                        q = q.filter(func.count(**i))
                    
                    if 'group_by' in i:
                        for k, v in i.items():
                            q = q.group_by(v)
                    
                    if fields['join']:
                        for i in fields['join']:
                            for k, v in i.items():
                                q = q.join(v)

                result = await s.execute(q)  # Now we have all elements of the query so we can execute it
                print ([row for row in result])  # Print out the query results

Hopefully this is clearer than my initial query!
Perhaps there are examples somewhere that I missed, I would be happy to be pointed in the right direction!

**Reply by CaselIT** [reply]
> For your answer where you give the example query.filter(sa.or_(sa.and_(table.c.id > 1, table.c.name != 'foo'), table.c.name=='bar')).
Is this using the ORM or Core?

that examples used core, but it also applied to the orm. It that case you would use `Mode.id` and `Model.name` instead of `table.c.id` and `table.c.name`.

the example you provided seems a usable starting point. Just a couple quick notes regarding `or_` and `where`: they accept sqlalchemy clauses, so if you have a column name you will want to use `getattr(Model, name) == value` there. A simple example is below

```py
import sqlalchemy as sa
from sqlalchemy.orm import declarative_base
Base = declarative_base()

class Model(Base):
    id = sa.Column(sa.Integer, primary_key=True)
    name = sa.Column(sa.Text)
    desc = sa.Column(sa.Text)
    __tablename__ = "model"

def do_where(model, query):
    if "or" in query:
        return sa.or_(*[do_where(model, and_) for and_ in query["or"]])
    else:
        return sa.and_(
            *[getattr(model, name) == value for name, value in query.items()]
        )

print(do_where(Model, {"name": "a", "desc": "b"}))
# model.name == 'a' AND model.desc == 'b'
print(do_where(Model, {"or": [{"name": "a", "desc": "b"}, {"name": "c"}]}))
# (model.name == 'a' AND model.desc == 'b') OR model.name == 'c'
```

Regarding other documentation on this, I don't think there a similar example in the sqlalchemy docs.

**Reply by fleetofthought** [reply]
Thanks for the examples @CaselIT !

Building on the starting point above, I am still not clear on how to build up a query properly. Here is my example using aggregates and `order by random`:

```
async def read(self, table, mods=None):
    Table = self._cache.classes[table]
    
    # Start the query object with simplest possible statement: select table. 'select()' on its own produces only empty queries
    q = select(Table)

    if mods and mods['fields']:
        for f in mods['fields']:
            if isinstance(f, dict):
                for k, v in f.items():
                    if k == 'avg':
                        q = q.filter(func.avg(getattr(Table, v)))
                    elif k == 'min':
                        q = q.filter(func.min(getattr(Table, v)))
                    elif k == 'max':
                        q = q.filter(func.max(getattr(Table, v)))
                    elif k == 'sum':
                        q = q.filter(func.sum(getattr(Table, v)))
                    elif k == 'count':
                        q = q.filter(func.count(getattr(Table, v)))
                    elif k == 'median':
                        q = q.filter(func.percentile_cont(0.5).within_group(getattr(Table, v)))
                    print(q)

    #Sorting
    if 'random' in mods:
        q = q.order_by(func.random())

db = DB(ENGINE, SYNC_ENGINE)

res = asyncio.run(DB.read('userapp',
                          query={'owner_id': '123a'},
                          mods={'fields':
                                    ['userapp.usage',
                                     {'max': 'allowance'},
                                     {'median': 'usage'},
                                     {'count': 'owner_id'},
                                     ],
                                'random': ['']}))
```

**Different Formats Tried**


I can use an aggregate function on its own like this:
`select(func.max(getattr(Table, v)))` which emits SQL `select max(appuser.allowance) from appuser;` Nice!

But this only works as a standalone. As soon as I put it in the format above:
```
q = select(Table) 
if k == 'max':
    q = q.filter(func.max(getattr(Table, v)))
```
If we print `q` at this point this is the invalid SQL emitted (Invalid because aggregate functions are not allowed in WHERE):
```
SELECT anon_1.id, anon_1.date_created, anon_1.owner_id, anon_1.allowance, anon_1.usage
FROM (SELECT appuser.id AS id, appuser.date_created AS date_created, appuser.owner_id AS owner_id, appuser.allowance AS allowance, appuser.usage AS usage
FROM appuser) AS anon_1, appuser 
WHERE max(appuser.allowance)
```

By the time we get to `    if 'random' in mods:` and print `q`, this is the SQL emitted:
```
SELECT anon_1.id, anon_1.date_created, anon_1.owner_id, anon_1.allowance, anon_1.usage
FROM (SELECT appuser.id AS id, appuser.date_created AS date_created, appuser.owner_id AS owner_id, appuser.allowance AS allowance, appuser.usage AS usage
FROM appuser) AS anon_1, appuser 
WHERE max(appuser.allowance) AND percentile_cont(:percentile_cont_1) WITHIN GROUP (ORDER BY appuser.usage) AND count(appuser.owner_id) ORDER BY random()
```

I have tried putting a blank `select` statement at the beginning of the function like `q  = select()` but this just results in empty select statements, the only SQL emitted is `SELECT`

I have tried to change how I start the query by doing things like:   `q = self._session()` But this results in errors like  `AttributeError: 'AsyncSession' object has no attribute 'select', AttributeError: 'AsyncSession' object has no attribute 'filter'`

If I remove `filter` from the aggregate queries like `q.func.max(getattr(Table, v))` results in error `AttributeError: 'Select' object has no attribute 'func'.

Trying to put `Table` directly into the aggregate queries like `q.filter(Table(func.max(getattr(Table, v))))` errors `TypeError: __init__() takes 1 positional argument but 2 were given`.

I thought maybe `filter` was incorrect to use here so I tried these variations which all error:
`q.filter_by(func.max(getattr(Table, v)))`
`TypeError: filter_by() takes 1 positional argument but 2 were given`

`q.values(func.max(getattr(Table, v)))`
`AttributeError: 'Select' object has no attribute 'value'`

I also tried building up the query like this but run into problems because of trying to call the `select()` statement:
```
if k == 'max':
      max_agg = func.max(getattr(Table, v))
       q1 = select(max_agg)  # prints SELECT max(appuser.allowance) AS max_1 FROM appuser - Nice!
       
       # But as soon as we try to initialise an empty `select` statement to "build into" for multiple if conditions, error:
       q2 = select()
       q2 = q2(max_agg)   # TypeError: 'Select' object is not callable
       q2 = q2.max_agg    #  AttributeError: 'Select' object has no attribute 'max_agg'
      q2 = q2.filter(max_agg)   # Original problem: Invalid SQL! SELECT  FROM appuser WHERE max(appuser.allowance)                     
```

I know it is possible to build up queries properly using this ORM so I think I must be overlooking some key functionality.

**Reply by CaselIT** [reply]
`filter` is used to add where clauses, not to add fields to the select list. To do that you would either create a list of fields beforehand and then create a select with `select(*fields)` or you could create and empty select and then add new columns with [`add_columns`](https://docs.sqlalchemy.org/en/14/core/selectable.html?highlight=add_column#sqlalchemy.sql.expression.Select.add_columns)

**Reply by fleetofthought** [reply]
Thank you _so _much__ for your helpful responses, they are really appreciated and are leading to greater SA understanding :)
